services/remotecluster/ping.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package remotecluster
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+// pingLoop periodically sends a ping to all remote clusters.
+func (rcs *Service) pingLoop(done <-chan struct{}) {
+	pingChan := make(chan *model.RemoteCluster, MaxConcurrentSends*2)
+
+	// create a thread pool to send pings concurrently to remotes.
+	for i := 0; i < MaxConcurrentSends; i++ {
+		go rcs.pingEmitter(pingChan, done)
+	}
+
+	go rcs.pingGenerator(pingChan, done)
+}
+
+func (rcs *Service) pingGenerator(pingChan chan *model.RemoteCluster, done <-chan struct{}) {
+	defer close(pingChan)
+
+	for {
+		start := time.Now()
+
+		// get all remotes, including any previously offline.
+		remotes, err := rcs.server.GetStore().RemoteCluster().GetAll(true)
+		if err != nil {
+			rcs.server.GetLogger().Log(mlog.LvlRemoteClusterServiceError, "Ping remote cluster failed (could not get list of remotes)", mlog.Err(err))
+			select {
+			case <-time.After(PingFreq):
+				continue
+			case <-done:
+				return
+			}
+		}
+
+		for _, rc := range remotes {
+			if rc.SiteURL != "" { // filter out unconfirmed invites
+				pingChan <- rc
+			}
+		}
+
+		// try to maintain frequency
+		elapsed := time.Since(start)
+		sleep := time.Second * 1
+		if elapsed < PingFreq {
+			sleep = time.Until(start.Add(PingFreq))
+		}
+
+		select {
+		case <-time.After(sleep):
+		case <-done:
+			return
+		}
+	}
+}
+
+// pingEmitter pulls Remotes from the ping queue (pingChan) and pings them.
+// Pinging a remote cannot take longer than PingTimeoutMillis.
+func (rcs *Service) pingEmitter(pingChan <-chan *model.RemoteCluster, done <-chan struct{}) {
+	for {
+		select {
+		case rc := <-pingChan:
+			if rc == nil {
+				break
+			}
```

Comment:
If `rc` is `nil` then it means `pingChan` was closed so we might as well return here?

Reply:
`rc==nil` could be channel closed, or a nil rc was enqueued. I don't want any chance of all remotes silently going `offline` just because a nil slipped in there.  We have a definitive exit from the loop already.

Reply:
Fair enough. The current code would panic before getting to enqueue a `nil` since we check for its `SiteURL` before sending it through the channel. If it's something unexpected maybe we should log its occurrence.

Reply:
Ok, I reviewed the code and you're right, the only way a nil comes out of that channel is if the channel is closed.  I changed `break` to `return` via https://github.com/mattermost/mattermost-server/pull/16432/commits/c2ecb318bd81bbacbbd01e57234d361ad5fb00a6